pkg/controller/admissionchecks/multikueue/remoteclient.go

```diff
@@ -0,0 +1,183 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multikueue
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+)
+
+type remoteController struct {
+	localClient   client.Client
+	watchCtx      context.Context
+	watchCancel   context.CancelFunc
+	remoteClients map[string]*remoteClient
+	wlUpdateCh    chan<- event.GenericEvent
+
+	// For unit testing only. There is now need of creating fully functional remote clients in the unit tests
+	// and creating valid kubeconfig content is not trivial.
+	// The full client creation and usage is validated in the integration and e2e tests.
+	restConfigFromKubeConfigOverride func([]byte) (*rest.Config, error)
+	newWithWatchOverride             func(config *rest.Config, options client.Options) (client.WithWatch, error)
+}
+
+func newRemoteController(watchCtx context.Context, localClient client.Client, wlUpdateCh chan<- event.GenericEvent) *remoteController {
+	watchCtx, watchCancel := context.WithCancel(watchCtx)
+	ret := &remoteController{
+		localClient: localClient,
+		watchCtx:    watchCtx,
+		watchCancel: watchCancel,
+		wlUpdateCh:  wlUpdateCh,
+	}
+	return ret
+}
+
+func (rc *remoteController) UpdateConfig(ctx context.Context, kubeConfigs map[string][]byte) error {
+	if rc.remoteClients == nil {
+		rc.remoteClients = make(map[string]*remoteClient, len(kubeConfigs))
+	}
+
+	for clusterName, c := range rc.remoteClients {
+		if kubeconfig, found := kubeConfigs[clusterName]; found {
+			if err := c.setConfig(kubeconfig); err != nil {
+				delete(rc.remoteClients, clusterName)
+				return err
+			}
+		} else {
+			c.watchCancel()
+			delete(rc.remoteClients, clusterName)
+		}
+	}
+	// create the missing ones
+	for clusterName, kubeconfig := range kubeConfigs {
+		if _, found := rc.remoteClients[clusterName]; !found {
+			c := newRemoteClient(rc.watchCtx, rc.localClient, rc.wlUpdateCh)
+
+			c.restConfigFromKubeConfigOverride = rc.restConfigFromKubeConfigOverride
+			c.newWithWatchOverride = rc.newWithWatchOverride
+
+			if err := c.setConfig(kubeconfig); err != nil {
+				return err
+			}
+			rc.remoteClients[clusterName] = c
+		}
+	}
+
+	return nil
+}
+
+func (cc *remoteController) IsActive() bool {
+	return cc != nil && len(cc.remoteClients) > 0
+}
+
+type remoteClient struct {
+	localClient  client.Client
+	client       client.WithWatch
+	wlUpdateCh   chan<- event.GenericEvent
+	rootWatchCtx context.Context
```

Comment:
You can pass the rootContext when necessary.

Reply:
When making the MutiKueueCluster a top level object we might be able only store the manager running context once.

Reply:
This seems doable even now, because `rootWatchCtx` is only used in `setConfig`, which is only used in `UpdateConfig`, called from `Reconcile`, where we have the `rootContext` at hand. So, we could propagate it. 

Also, I find it  confusing that the root context is named differently at different places, `rootContext`, `rootWatchCtx`, or `watchCtx`. Can we align to call it just `rootContext`?

Reply:
Do you mean about adding the "cancel" in [here](https://github.com/kubernetes-sigs/kueue/blob/3b847e34603cd02b91008966ac67e080a3d20068/pkg/controller/admissionchecks/multikueue/remoteclient.go#L48)? 

Any reason we could not move it down as the first line in `UpdateConfig`? IIUC we could pass the derived context to `setConfig`.

Reply:
done